### ISSUE TITLE
Merge of CLI lib just breaks the CI  test.

### ISSUE THREAD

**Issue by shwsun** [CCI-MOC/hil#759]
#617 breaks the CI test we have, see https://travis-ci.org/CCI-MOC/hil/builds/225312311. @zenhack @henn @SahilTikale @naved001 

PS, the time to run CI is around 20min now.

**Comment by SahilTikale** [comment]
I am aware that the tests take longer since the unit tests require a server to be spinned up and sometimes it fails to fetch the given port. 
I have talked to @zenhack  about refactoring the tests so that we can run them in parallel and reduce the time of testing. Will do that once we close on the current PR and move from haas to hil

**Comment by naved001** [comment]
@shwsun the build you linked seems to be still running the tests (it passed for postgres)

> sometimes it fails to fetch the given port.

@SahilTikale does that mean that the tests might fail sometimes? it looks like it's passing at the moment.

**Comment by SahilTikale** [comment]
Yup that is what I am saying. It sometimes passes for one database and may fail for other, just need to restart that build. No changes required. 

About the tests taking long time, that is because it spins up a new server and tests against it, It makes rest calls for each test, that might take time.

**Comment by henn** [comment]
@SahilTikale Are you saying these tests will fail intermittently through
no fault of the potential submitter?

If so, that's something that would need to get fixed ASAP if we want to
keep this PR so as to not impact others.

**Comment by zenhack** [comment]
Re: the tests taking a long time, I'm working on an improvement for that right now.

Re: intermittent failures, yeah, that's absolutely unacceptable.

**Comment by zenhack** [comment]
So I've got some WIP here:

https://github.com/zenhack/hil/tree/ci-speedup

The failures are in the keystone tests, but the client tests in particular are taking an insanely long time. I am debugging locally.

**Comment by SahilTikale** [comment]
@zenhack I have figured out why it was taking this long. 
Working on a patch.

**Comment by SahilTikale** [comment]
I may be able to get it down to less than a minute.

**Comment by SahilTikale** [comment]
I have got the patch ready. Need to clean up a little bit. will make a new PR in 2 hours.

**Comment by zenhack** [comment]
Ok, cool.

**Comment by zenhack** [comment]
Any notions about the keystone tests? I'm a bit puzzled by those, since I got the same failure when I separated out the keystone tests from the others, so it's not really being touched by your stuff (I think), but it's still passing on the old master, so it's not that something changed out from under us either.

**Comment by knikolla** [comment]
Why are these in unit tests? We have a full blown apache apache for our integration tests (which these are), use that instead of spinning up a new server. These are not unit tests.

**Comment by SahilTikale** [comment]
@zenhack I did all tests on my local branch, got finished at a good pace. 
I have made an urgent patch for the client library tests. This should help shave off some time from the duration it took to run the tests in travis.

**Comment by shwsun** [comment]
> @shwsun the build you linked seems to be still running the tests (it passed for postgres)

Yes, I was a bit nervous about the merge yesterday and open this issue as soon as the sqlite test fail.


> Yup that is what I am saying. It sometimes passes for one database and may fail for other, just need to restart that build. No changes required.

So the long time running CI is fine to me, Openstack developers have to deal with it all the time. I am just complaining that it (maybe) should not be that long for a project at HIL's scale. 

I agree with @zenhack and @henn , one thing really nervous me is that now sqlite or postgres tests may fail for no reason. It probably means some part of the CLI lib screws our testing infrastructure, and it should be addressed soon.

**Comment by SahilTikale** [comment]
I removed client.py from the unit tests and rerun the tests. It still failed with the following error

```
(.venv)sahil@rHaaS:~/git/HIL/tests/unit$ py.test .
========================================================================================================= test session starts =========================================================================================================
platform linux2 -- Python 2.7.6, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/sahil/git/HIL, inifile: setup.cfg
plugins: xdist-1.14, cov-1.8.0
collected 462 items / 1 errors 
class_resolver.py ...
cli.py .....
config.py .
dev_support.py ....
haas_auth.py .
migrations.py ss
model.py ................
test_common.py ....................................
api/auth.py ...................................................................................................................................................................................
api/main.py ........................................................................................................................................................................
api/port_revert.py ...
ext/auth/database.py ...............................
ext/auth/mock.py .....
ext/network_allocators/vlan_pool.py ..
ext/obm/ipmi.py ..
ext/switches/brocade.py ....

================= ERRORS ==============================================
__________________ ERROR collecting tests/unit/rest.py ____________________________

```

Truncating output, getting to the error directly

```
../../.venv/local/lib/python2.7/site-packages/_pytest/config.py:392: in import_plugin
    raise new_exc
E   ImportError: Error importing plugin "pytest_catchlog": No module named pytest_catchlog
================ 460 passed, 2 skipped, 1 error in 60.67 seconds =====================

```

This error got resolved when I installed the said package. `pytest_catchlog`
I am working on the latest master branch. Did anyone remove the package during one of the prior merges ?

```
(.venv)sahil@rHaaS:~/git/HIL/tests/unit$ pip install pytest-catchlog
Downloading/unpacking pytest-catchlog
  Downloading pytest_catchlog-1.2.2-py2.py3-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): pytest>=2.6 in /home/sahil/git/HIL/.venv/lib/python2.7/site-packages (from pytest-catchlog)
Requirement already satisfied (use --upgrade to upgrade): py>=1.1.1 in /home/sahil/git/HIL/.venv/lib/python2.7/site-packages (from pytest-catchlog)
Installing collected packages: pytest-catchlog
Successfully installed pytest-catchlog
Cleaning up...
(.venv)sahil@rHaaS:~/git/HIL/tests/unit$ 
(.venv)sahil@rHaaS:~/git/HIL/tests/unit$ 
(.venv)sahil@rHaaS:~/git/HIL/tests/unit$ py.test .
================================================================== test session starts ===========================================================================
platform linux2 -- Python 2.7.6, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/sahil/git/HIL, inifile: setup.cfg
plugins: xdist-1.14, cov-1.8.0, catchlog-1.2.2
collected 499 items 

class_resolver.py ...
cli.py .....
config.py .
dev_support.py ....
haas_auth.py .
migrations.py ss
model.py ................
rest.py .....................................
test_common.py ....................................
api/auth.py ...................................................................................................................................................................................
api/main.py ........................................................................................................................................................................
api/port_revert.py ...
ext/auth/database.py ...............................
ext/auth/mock.py .....
ext/network_allocators/vlan_pool.py ..
ext/obm/ipmi.py ..
ext/switches/brocade.py ....

=========== 497 passed, 2 skipped in 61.34 seconds ==========================
```

**Comment by zenhack** [comment]
@SahilTikale, that dependency was introduced recently; I suspect you were working from a venv you created prior to that patch being merged?

**Comment by shwsun** [comment]
@SahilTikale   It is introduced by #756

**Comment by naved001** [comment]
We can close this issue now, since the CI is stable after we merged #774 

To discuss performance issues, we have #764 open
